### PR TITLE
Enable env variables

### PR DESCRIPTION
**Pull request by kepta**
This enables the use of `.env` files following the algorithm [here](https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use). This works similar to how [create-react-app](https://github.com/facebook/create-react-app/blob/master/packages/react-scripts/template/README.md#adding-development-environment-variables-in-env) handles env vars.

This also adds the option of passing `mode` option via CLI, this would either be `development` or `production`. The current problem I am facing when running `npx underreact start -m=production` is that the `html-compiler` is not correctly able to serve CSS passed through `underreact.config.js`.

I feel the problem lies somewhere [here](https://github.com/mapbox/underreact/blob/b14fed5edf90cab492280ed046d37524511efb1e/lib/html-compiler.js#L30). In the `start` command [this function](https://github.com/mapbox/underreact/blob/b14fed5edf90cab492280ed046d37524511efb1e/lib/html-compiler.js#L13) is not called with `cssFilename` and [cssAssets](https://github.com/mapbox/underreact/blob/b14fed5edf90cab492280ed046d37524511efb1e/lib/html-compiler.js#L41) ends up missing the file compiled by postCSS plugin. Thoughts?

Will add documentation regarding this in the next PR.

@davidtheclark for review, please.

### CODE REVIEW

File: lib/utils/parse-argv.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
+'use strict';
+const path = require('path');
+
+module.exports = function parseArgv(command, argv = {}) {
```
Comment (davidtheclark) [inline lib/utils/parse-argv.js]: argv has already been parsed, right? So maybe there's a name that better describes what this function does?

File: lib/utils/set-env.js [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+  const dontenvPath = path.join(rootDirectory, '.env');
+
+  // https://github.com/bkeepers/dotenv#what-other-env-files-can-i-use
+  const dotenvFiles = [
+    NODE_ENV && `${dontenvPath}.${NODE_ENV}.local`,
+    NODE_ENV && `${dontenvPath}.${NODE_ENV}`,
+    // Don't include `.env.local` for `test` environment
+    // since normally you expect tests to produce the same
+    // results for everyone
+    NODE_ENV !== 'test' && `${dontenvPath}.local`,
```
Comment (davidtheclark) [inline lib/utils/set-env.js]: Do we have any use cases for a `.local` file? If not, I'd be in favor of removing the code that makes it work.

File: lib/config/index.js [hunk 3]
```diff
@@ -1,68 +1,23 @@
...
-      const defaultConfig = getDefaultConfig(
-        Object.assign({}, configModuleContext, {
-          argv,
-          configDir: path.dirname(configPath),
-          stats: argv.stats,
-          port: argv.port
-        })
-      );
-      return createUrc(userConfig, defaultConfig);
+function config(cliOpts) {
```
Comment (davidtheclark) [inline lib/config/index.js]: Could you add a comment above this function describing what it does? (What it accepts, what it returns?)

File: lib/config/index.js [hunk 4]
```diff
@@ -1,68 +1,23 @@
...
+  return userConfigReader(cliOpts.configPath).then(getUserConfig => {
```
Comment (davidtheclark) [inline lib/config/index.js]: Could `userConfigReader` be renamed to `getUserConfig` and accept the other options (`webpack, command, mode`)? If so we can avoid an additional higher-order function that makes the code a little harder to understand.

File: lib/config/create-urc.js [hunk 5]
```diff
@@ -38,6 +38,30 @@ class Urc {
       deleteCache: true
     });
   }
+
+  getClientEnvVars() {
+    // Inspired from https://github.com/facebook/create-react-app/blob/next/packages/react-scripts/config/env.js#L71
+    // Grabs NODE_ENV and environment variables that startWith CLIENT_ENV_PREFIX and prepare them to be
+    // injected into the application via EnvironmentPlugin in Webpack configuration.
```
Comment (davidtheclark) [inline lib/config/create-urc.js]: I see I reviewed a couple of PRs in the wrong order —— see my comments on the other PR for my reservations about requiring a prefix for these env variables.

File: commands/start.js [hunk 6]
```diff
@@ -24,15 +20,17 @@ const {
...
+  );
   return (
     del(urc.outputDirectory, { force: true })
       // webpack needs to run first as others depend on the assets
       .then(() => new Promise(resolve => watchWebpack(urc, resolve)))
       .then(() => {
         Promise.all([
           writeHtml(urc),
-          writeCss(urc),
+          writeCss(urc).then(r => console.log(r)),
```
Comment (davidtheclark) [inline commands/start.js]: Is this leftover from debugging?

File: commands/start.js [hunk 7]
```diff
@@ -104,7 +102,9 @@ function watchCss(urc) {
   });
 
   watchStyleSheets.on('all', () => {
-    writeCss(urc).catch(logger.error);
+    writeCss(urc)
+      .then(r => console.log(r))
```
Comment (davidtheclark) [inline commands/start.js]: Is this leftover from debugging?

File: bin/underreact.js [hunk 8]
```diff
@@ -106,12 +147,16 @@ function defineServeStatic(y) {
 }
 
 function runServeStatic(argv) {
-  config('serve-static', argv)
+  const serveStatic = require('../commands/serve-static');
+  const config = require('../lib/config');
```
Comment (davidtheclark) [inline bin/underreact.js]: If every command ends up requiring `config`, why not require it at the top of the file?

Comment (kepta) [reply bin/underreact.js]: The problem with requiring it at the top of file would mean that it wouldn't get the necessary `env` variable setup, leading to unpredictability. Hence I am trying to defer most of the `require`s until the [middleware](https://github.com/mapbox/underreact/pull/21/files#diff-ecf66e53d7f5563584528930a084b7f4R15) has executed, in order to setup the necessary `env` vars.

Comment (kepta) [reply bin/underreact.js]: We would want our env variables to be set before anything else executes and to make that happen we also need a normalised `argv` coming from `yargs`, that is why you would the env vars are being set inside the yargs middleware. I have left a comment about that [here](https://github.com/mapbox/underreact/blob/b14fed5edf90cab492280ed046d37524511efb1e/bin/underreact.js#L10)

File: commands/start.js (hunk 7 above)
Comment (kepta) [reply commands/start.js]: Yup sorry about that.

**Comment by kepta** [comment]
@davidtheclark updated the PR with fixes.

File: bin/underreact.js (hunk 8 above)
Comment (davidtheclark) [reply bin/underreact.js]: Oh, huh 👍

**Pull request merged**